### PR TITLE
Add saga prototype

### PR DESCRIPTION
**Pull request by drets**

### CODE REVIEW

File: framework/src/main/java/com/eventor/internal/ClassProcessor.java [hunk 1]
```diff
@@ -44,10 +47,18 @@ private MetaSubscriber handleSubscriber(Class<?> clazz) {
 
     private MetaAggregate handleAggregate(Class<?> aggregateClass) {
         Map<Class<? extends Annotation>, Iterable<Method>> annotated =
-                EventorReflections.getMethodsAnnotated(aggregateClass, EventHandler.class, CommandHandler.class, EventListener.class);
```
Comment (st-kurilin) [inline framework/src/main/java/com/eventor/internal/ClassProcessor.java:47]: Now we handle both EventHandler AND EventListener. In future we will remove EventHandler. EventListener already in this list.

File: framework/src/test/groovy/com/eventor/university/Course.java [hunk 2]
```diff
@@ -36,6 +37,15 @@ public Object on(@IdIn("courseId") SubmitResult cmd) {
         }
     }
 
+    @CommandHandler
+    public Object on(@IdIn("courseId") StartCourseExam cmd) {
+        if (attempt) {
+            throw new RuntimeException("Only one attempt for course exam");
+        }
+        attempt = true;
```
Comment (st-kurilin) [inline framework/src/test/groovy/com/eventor/university/Course.java]: Don't make changes in Command handler. add event handler on CourseExamStarted and change state there.

File: framework/src/test/groovy/com/eventor/university/CourseExamStarted.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.eventor.university;
+
+public class CourseExamStarted {
```
Comment (st-kurilin) [inline framework/src/test/groovy/com/eventor/university/CourseExamStarted.java]: Let's name it CourseFinalExam it smt like that.

**Comment by st-kurilin** [comment]
Make all fields public final or private final or private.

File: framework/src/main/java/com/eventor/internal/ClassProcessor.java (hunk 1 above)
Comment (drets) [inline framework/src/main/java/com/eventor/internal/ClassProcessor.java:47]: thanks, for now, I can see.

File: framework/src/main/java/com/eventor/internal/Akka.java [hunk 4]
```diff
@@ -27,6 +29,11 @@ public Invokable createBroadcaster(Iterable<Listener> listeners) {
         return wrap(system.actorOf(Props.empty().withRouter(BroadcastRouter.create(refs))));
     }
 
+    public Cancellable createScheduler(FiniteDuration duration, MetaSubscriber subscriber, Object args, ActorRef sender) {
```
Comment (st-kurilin) [inline framework/src/main/java/com/eventor/internal/Akka.java]: Do we need it now?

File: framework/src/test/groovy/com/eventor/EventorTest.groovy [hunk 5]
```diff
@@ -56,8 +56,18 @@ class EventorTest extends Specification {
         instanceCreator.getInstanceOf(Grades).grade("Math-2013", "Poll") == 120
     }
 
+    def "Saga has to receive events and generate commands"() {
+        when:
+        ch.registerCourse("Math")
+        cb.submit(new StartCourseExam("Math", "Alica")); // Alica is daughter of the boss :)
```
Comment (st-kurilin) [inline framework/src/test/groovy/com/eventor/EventorTest.groovy]: Let's do it without jokes for now)

File: framework/src/main/java/com/eventor/Eventor.java [hunk 6]
```diff
@@ -140,6 +144,29 @@ private void handleEventByAggregate(Object aggregate, MetaHandler eachMetaHandle
...
+        for (MetaHandler eachMetaHandler : eachMetaSaga.eventHandlers) {
+            if (eachMetaHandler.expected.equals(event.getClass())) {
+                Collection<?> messages = eachMetaHandler.execute(instanceCreator.getInstanceOf(eachMetaSaga.origClass), event);
+                Cancellable scheduler = null;
+                for (Object message : messages) {
+                    if (message instanceof Finish) {
+                        if (scheduler != null) {
+                            scheduler.cancel();
+                        }
+                    }
```
Comment (st-kurilin) [inline framework/src/main/java/com/eventor/Eventor.java]: Please don't start new line before else

File: framework/src/main/java/com/eventor/Eventor.java [hunk 7]
```diff
@@ -97,6 +100,21 @@ private void handleCmdByAggregate(MetaAggregate eachMetaAggregate, Object cmd) {
         }
     }
 
+    private void handleCmdBySaga(MetaSaga eachMetaSaga, Object cmd) {
+        for (MetaHandler eachMetaHandler : eachMetaSaga.commandHandlers) {
+            if (eachMetaHandler.expected.equals(cmd.getClass())) {
+                Object sagaId = eachMetaHandler.extractId(cmd);
+                if (sagas.containsKey(sagaId)) {
+                    Object saga = sagas.get(sagaId);
+                    Collection<?> messages = EventorCollections.toCollection(eachMetaHandler.execute(saga, cmd));
```
Comment (st-kurilin) [inline framework/src/main/java/com/eventor/Eventor.java]: messages?

File: framework/src/main/java/com/eventor/Eventor.java [hunk 8]
```diff
@@ -140,6 +158,32 @@ private void handleEventByAggregate(Object aggregate, MetaHandler eachMetaHandle
...
+                        handleEventBySaga(saga, eachMetaHandler, event);
+                    }
+                }
+            }
+        }
+    }
+
+    private void handleEventBySaga(Object aggregate, MetaHandler eachMetaHandler, Object event) {
+        Collection<?> res = eachMetaHandler.execute(aggregate, event);
+        if (!res.isEmpty()) throw new RuntimeException("Void expected");
```
Comment (st-kurilin) [inline framework/src/main/java/com/eventor/Eventor.java]: Why void expected?

File: framework/src/test/groovy/com/eventor/university/Course.java [hunk 9]
```diff
@@ -36,7 +37,25 @@ public Object on(@IdIn("courseId") SubmitResult cmd) {
...
+    public Object on(@IdIn("courseId") GraduateExam cmd) {
+        int rightAnswers = 0;
+        for (int[] answer : cmd.answers) {
+            for (int value : answer) {
+                if (value == 42) {
+                    rightAnswers++;
+                }
+            }
+        }
+        return new FinalExamSubmitted(cmd.courseId, cmd.studentId, rightAnswers * 20);
```
Comment (st-kurilin) [inline framework/src/test/groovy/com/eventor/university/Course.java]: I believe grade should be between 0 and 100

File: framework/src/test/groovy/com/eventor/university/ExamGraduating.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    @Start
+    public void on(FinalExamStarted evt) {
+        id = evt.courseId;
+        answers = new int[evt.parts][];
+    }
+
+    @CommandHandler
+    public Object on(@IdIn("courseId") SubmitPartAnswer cmd) {
+        answers[cmd.part - 1] = cmd.answer;
+        if (answers.length == cmd.part) {
```
Comment (st-kurilin) [inline framework/src/test/groovy/com/eventor/university/ExamGraduating.java]: always true?

File: framework/src/test/groovy/com/eventor/university/FailedOnCourseResultsSubmission.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.eventor.university;
+
+public class FailedOnCourseResultsSubmission {
+    final String courseId;
```
Comment (st-kurilin) [inline framework/src/test/groovy/com/eventor/university/FailedOnCourseResultsSubmission.java]: public?

File: framework/src/main/java/com/eventor/api/annotations/Saga.java [hunk 12]
```diff
@@ -1,4 +1,10 @@
...
 
+import java.lang.annotation.*;
+
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Inherited
 public @interface Saga {
-}
+}
```
Comment (st-kurilin) [inline framework/src/main/java/com/eventor/api/annotations/Saga.java:10]: ?

File: framework/src/test/groovy/com/eventor/university/ExamGraduating.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+            for (int i = 0; i < answers.length; i++) {
+                answers[i] = submission.get(i + 1);
+            }
+            return new GraduateExam(cmd.courseId, cmd.studentId, answers);
+        } else {
+            return null;
+        }
+    }
+
+    private boolean checkPart(int part) {
```
Comment (st-kurilin) [inline framework/src/test/groovy/com/eventor/university/ExamGraduating.java]: lets do it more redable, eg
if (part < 0 || part >= parts) thr IllegalArgument
if (submission.keySet().contains(part)) thr IllegalState

**Pull request merged**